### PR TITLE
Add local build status

### PR DESCRIPTION
**Pull request by Meklo**

### CODE REVIEW

**Review by flomillot — APPROVED** [review]

**Comment by SlimaneAmar** [comment]
I think there is no need to store global status in the BD and it can be inferred automatically ?

**Comment by Meklo** [comment]
> I think there is no need to store global status in the BD and it can be inferred automatically ?

It implies that we would have to retrieve a node's parent each time there's a request for its informations.
Wouldn't it significantly increase the number of access to the database since retrieving a node's data is a recurrent operation ?

**Comment by SlimaneAmar** [comment]
Yes i think we need to optimize database access

**Comment by SlimaneAmar** [comment]
The 2 states are intimately linked so it is better to always have a method that handles both
Otherwise we risk forgetting one of the two

**Comment by Meklo** [comment]
> The 2 states are intimately linked so it is better to always have a method that handles both Otherwise we risk forgetting one of the two

Yes I agree, that is why I updated updateBuildStatus parameters instead of creating a new function 
[see here](https://github.com/gridsuite/study-server/pull/379/files#diff-7bcb8b769a3e7bd9dabfb88cf01a0e5585d36505ab547f06e2cf633616069b0eR115)
and invalidateBuildStatus also handles both [see here](https://github.com/gridsuite/study-server/pull/379/files#diff-5b97a2d8d9e8aafc4b17cdbc97a7d4c7be71f96c3129570a9efedb19e16d1351R243)
Do you see another method where it could generate a desynchronization of the two ?

File: src/main/java/org/gridsuite/study/server/dto/modification/NetworkModificationResult.java [hunk 1]
```diff
@@ -10,9 +10,7 @@
 import io.swagger.v3.oas.annotations.media.Schema;
 import lombok.*;
 
-import java.util.List;
-import java.util.Set;
-import java.util.TreeSet;
+import java.util.*;
```
Comment (Walid-Sahnoun) [inline src/main/java/org/gridsuite/study/server/dto/modification/NetworkModificationResult.java:13]: no need to import all library

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java [hunk 2]
```diff
@@ -232,12 +243,16 @@ public UUID getVoltageInitResultUuid(UUID nodeUuid) {
         return getVoltageInitResultUuid(getNode(nodeUuid));
     }
 
-    public void updateBuildStatus(UUID nodeUuid, BuildStatus buildStatus, List<UUID> changedNodes) {
-        updateBuildStatus(getNode(nodeUuid), buildStatus, changedNodes);
+    public void updateBuildStatus(UUID nodeUuid, BuildStatus buildStatusGlobal, BuildStatus buildStatusLocal, List<UUID> changedNodes) {
+        updateBuildStatus(getNode(nodeUuid), buildStatusGlobal, buildStatusLocal, changedNodes);
+    }
+
+    public BuildStatus getBuildStatusGlobal(UUID nodeUuid) {
```
Comment (AbdelHedhili) [inline src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java:250]: getBuildStatusGlobal -> getGlobalBuildstatus
getBuildStatusLocal -> getLocalBuildStatus ?

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/NodeBuildStatus.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package org.gridsuite.study.server.networkmodificationtree.dto;
```
Comment (AbdelHedhili) [inline src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/NodeBuildStatus.java:1]: copyrights

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/NodeBuildStatus.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+import lombok.*;
+import lombok.experimental.SuperBuilder;
+
+@SuperBuilder
+@Getter
+@AllArgsConstructor
+@NoArgsConstructor
+@Setter
+@EqualsAndHashCode
+public class NodeBuildStatus {
```
Comment (AbdelHedhili) [inline src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/NodeBuildStatus.java:13]: author

**Review by AbdelHedhili — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/study/server/dto/modification/NetworkModificationResult.java (hunk 1 above)
Comment (Meklo) [reply src/main/java/org/gridsuite/study/server/dto/modification/NetworkModificationResult.java]: Done

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java (hunk 2 above)
Comment (Meklo) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java]: Fine by me, I've done the change across the PR

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/NodeBuildStatus.java (hunk 3 above)
Comment (Meklo) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/NodeBuildStatus.java:7]: Done

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/NodeBuildStatus.java (hunk 4 above)
Comment (Meklo) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/dto/NodeBuildStatus.java:22]: Done

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java [hunk 5]
```diff
@@ -57,7 +58,11 @@ public LoadFlowInfos getLoadFlowInfos(AbstractNode node) {
         return LoadFlowInfos.builder().loadFlowStatus(LoadFlowStatus.NOT_DONE).build();
     }
 
-    public BuildStatus getBuildStatus(AbstractNode node) {
+    public BuildStatus getGlobalBuildStatus(AbstractNode node) {
```
Comment (AbdelHedhili) [inline src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java:61]: parameter unused ?

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java [hunk 6]
```diff
@@ -57,7 +58,11 @@ public LoadFlowInfos getLoadFlowInfos(AbstractNode node) {
...
+        return BuildStatus.NOT_BUILT;
+    }
+
+    public BuildStatus getLocalBuildStatus(AbstractNode node) {
         return BuildStatus.NOT_BUILT;
```
Comment (AbdelHedhili) [inline src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java:66]: parameter unused ?

**Review by AbdelHedhili — CHANGES_REQUESTED** [review]

File: src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java (hunk 6 above)
Comment (AbdelHedhili) [reply src/main/java/org/gridsuite/study/server/networkmodificationtree/AbstractNodeRepositoryProxy.java:66]: should be abstract

**Review by AbdelHedhili — CHANGES_REQUESTED** [review]

**Review by AbdelHedhili — APPROVED** [review]

**Review by Walid-Sahnoun — APPROVED** [review]
code ok
test ok

**Review by SlimaneAmar — APPROVED** [review]

**Pull request merged**